### PR TITLE
add runtime architecture to RA and RD

### PR DESCRIPTION
**Pull request by nilsgstrabo**
- Added Runtime struct to RadixApplication components and jobs in common and environment config. Runtime has one prop, Architecture, which can be amd64 (the default) or arm64
- Configure job and component affinity based on Runtime in RD
- Building arm64 is only supported for BuildKit enabled applications. Validation is performed in pipeline, and will print an error if using arm64 without BuildKit. The build steps are run on respectitve architecture defined for the component or job. 
- Restructured pipeline steps into separate packages

Also deleted BoolPtr in radix-operator with pointers.Ptr from radix-common, and removed some unused code

### CODE REVIEW

File: pipeline-runner/steps/applyconfig/step.go [hunk 1]
```diff
@@ -146,15 +143,40 @@ func (cli *ApplyConfigStepImplementation) setBuildAndDeployImages(ctx context.Co
...
+			hasNonDefaultRuntimeArchitecture = slice.Any(buildComponents, func(c pipeline.BuildComponentImage) bool {
+				return operatorutils.GetArchitectureFromRuntime(c.Runtime) != defaults.DefaultNodeSelectorArchitecture
+			})
+			if hasNonDefaultRuntimeArchitecture {
+				break
+			}
+		}
+		if hasNonDefaultRuntimeArchitecture {
+			return ErrBuildNonDefaultRuntimeArchitectureWithoutBuildKitError
+		}
```
Comment (satr) [inline pipeline-runner/steps/applyconfig/step.go:174]: It can be simplified with if ... return

Comment (nilsgstrabo) [reply pipeline-runner/steps/applyconfig/step.go]: Tried to make it more readable.  Please check again.

File: pipeline-runner/steps/promote/errors.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+func EmptyArgument(argumentName string) error {
+	return fmt.Errorf("%s cannot be empty", argumentName)
+}
+
+// NonExistingFromEnvironment From environment does not exist
+func NonExistingFromEnvironment(environment string) error {
+	return fmt.Errorf("non existing from environment %s", environment)
+}
+
+// NonExistingToEnvironment From environment does not exist
```
Comment (Richard87) [inline pipeline-runner/steps/promote/errors.go:15]:
```suggestion
// NonExistingToEnvironment To environment does not exist
```

File: pipeline-runner/steps/build/build_acr.go [hunk 3]
```diff
@@ -51,7 +53,7 @@ func (step *BuildStepImplementation) buildContainerImageBuildingJobsForACRTasks(
 
 	log.Debug().Msg("build a build-job")
 	hash := strings.ToLower(utils.RandStringStrSeed(5, pipelineInfo.PipelineArguments.JobName))
-	job := buildContainerImageBuildingJob(rr, pipelineInfo, buildSecrets, hash, buildComponentImages...)
+	job := buildContainerImageBuildingJob(rr, pipelineInfo, buildSecrets, hash, nil, buildComponentImages...)
```
Comment (satr) [inline pipeline-runner/steps/build/build_acr.go:56]: Send explicit runtime?

File: pipeline-runner/internal/tekton/tekton.go [hunk 4]
```diff
@@ -49,21 +51,29 @@ func CreateActionPipelineJob(containerName string, action string, pipelineInfo *
...
+								securitycontext.WithContainerSeccompProfileType(corev1.SeccompProfileTypeRuntimeDefault),
+								securitycontext.WithReadOnlyRootFileSystem(pointers.Ptr(true)),
+							),
+							Env: *envVars,
 						},
 					},
 					Volumes:       getJobVolumes(),
 					RestartPolicy: "Never",
-					Affinity:      utils.GetAffinityForPipelineJob(),
+					Affinity:      utils.GetAffinityForPipelineJob(nil),
```
Comment (satr) [inline pipeline-runner/internal/tekton/tekton.go:76]: Send explicit runtime?

File: pkg/apis/job/kubejob.go [hunk 5]
```diff
@@ -109,7 +109,7 @@ func (job *Job) getPipelineJobConfig(ctx context.Context) (*batchv1.Job, error)
 						},
 					},
 					RestartPolicy: "Never",
-					Affinity:      utils.GetAffinityForPipelineJob(),
+					Affinity:      utils.GetAffinityForPipelineJob(nil),
```
Comment (satr) [inline pkg/apis/job/kubejob.go:112]: Send explicit runtime?

File: pkg/apis/deployment/jobschedulercomponent.go [hunk 6]
```diff
@@ -85,6 +85,10 @@ func (js *jobSchedulerComponent) GetNode() *v1.RadixNode {
 	return nil
 }
 
+func (js *jobSchedulerComponent) GetRuntime() *v1.Runtime {
+	return nil
```
Comment (satr) [inline pkg/apis/deployment/jobschedulercomponent.go:89]: Return runtime explicitly?

**Review by Richard87 — CHANGES_REQUESTED** [review]
tinye comments

File: pkg/apis/utils/affinity.go [hunk 7]
```diff
@@ -42,7 +42,7 @@ func GetAffinityForOAuthAuxComponent() *corev1.Affinity {
 			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
 				NodeSelectorTerms: []corev1.NodeSelectorTerm{
 					{
-						MatchExpressions: getNodeSelectorRequirementsForRuntimeEnvironment(),
+						MatchExpressions: getNodeSelectorRequirementsForRuntimeEnvironment(nil),
```
Comment (satr) [inline pkg/apis/utils/affinity.go:45]: Explicitly define runtime?

File: pkg/apis/utils/affinity.go [hunk 8]
```diff
@@ -56,50 +56,50 @@ func GetAffinityForJobAPIAuxComponent() *corev1.Affinity {
...
```
Comment (satr) [inline pkg/apis/utils/affinity.go:59]: Explicitly define runtime?

File: pipeline-runner/steps/build/build_acr.go (hunk 3 above)
Comment (nilsgstrabo) [reply pipeline-runner/steps/build/build_acr.go]: done

File: pipeline-runner/internal/tekton/tekton.go (hunk 4 above)
Comment (nilsgstrabo) [reply pipeline-runner/internal/tekton/tekton.go]: done

File: pkg/apis/job/kubejob.go (hunk 5 above)
Comment (nilsgstrabo) [reply pkg/apis/job/kubejob.go]: done

File: pkg/apis/utils/affinity.go (hunk 7 above)
Comment (nilsgstrabo) [reply pkg/apis/utils/affinity.go]: done

File: pkg/apis/utils/affinity.go (hunk 8 above)
Comment (nilsgstrabo) [reply pkg/apis/utils/affinity.go]: done

**Review by Richard87 — APPROVED** [review]

**Pull request merged**